includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -999,19 +1002,54 @@ private function get_disallowed_attributes_in_node( DOMElement $node, $attr_spec
 		return $attrs_to_remove;
 	}
 
+	/**
+	 * Adapt width and height to conform to the AMP layout.
+	 *
+	 * @param DOMElement $node Node.
+	 */
+	private function adapt_dimensions_to_layout( DOMElement $node ) {
+		$layout = $node->getAttribute( 'layout' );
+
+		if ( ! amp_is_attribute_empty( $layout ) ) {
+			$width  = $node->getAttribute( 'width' );
+			$height = $node->getAttribute( 'height' );
+
+			// If the width & height are `100%` the layout must be `fill`.
+			if ( '100%' === $width && '100%' === $height ) {
+				$node->removeAttribute( 'width' );
+				$node->removeAttribute( 'height' );
+				$node->setAttribute( 'layout', AMP_Rule_Spec::LAYOUT_FILL );
+				return;
+			}
+
+			// If the width is `100%`, convert the layout to `fixed-height` and width to `auto`.
+			if ( '100%' === $width ) {
+				$node->setAttribute( 'width', 'auto' );
+				$node->setAttribute( 'layout', AMP_Rule_Spec::LAYOUT_FIXED_HEIGHT );
+			}
+		}
+	}
+
 	/**
 	 * Remove invalid AMP attributes values from $node that have been implicitly disallowed.
 	 *
 	 * Allowed values are found $this->globally_allowed_attributes and in parameter $attr_spec_list
 	 *
 	 * @param DOMElement $node                       Node.
+	 * @param array[][]  $tag_spec                   Tag spec list.
 	 * @param array[][]  $attr_spec_list             Attribute spec list.
 	 * @param DOMAttr[]  $attributes_pending_removal Attributes pending removal.
 	 * @return DOMAttr[]|false Attributes to remove, or false if the element itself should be removed.
 	 */
-	private function sanitize_disallowed_attribute_values_in_node( DOMElement $node, $attr_spec_list, $attributes_pending_removal ) {
+	private function sanitize_disallowed_attribute_values_in_node( DOMElement $node, $tag_spec, $attr_spec_list, $attributes_pending_removal ) {
 		$attrs_to_remove = [];
 
+		if ( ! $this->is_valid_layout( $tag_spec, $node ) ) {
+			// If the layout is invalid, the node should be removed.
+			$this->remove_node( $node );
```

Comment:
Should the node be removed or just the _attribute_ be removed (adding `layout` attribute to `$attrs_to_remove`)? It should only be removed if `layout` is mandatory, per below.

Reply:
I'd say it should be removed. For example, an `amp-img` with `width=""` and `layout="responsive"` has a valid width and layout on its own, but because the layout is not mandatory, it will be considered valid markup.

Reply:
I don't understand. Given markup like this:

```
<amp-img src="https://example.com/foo.jpg" width="100" height="200" layout="bogus"></amp-img>
```

The `layout` attribute is not valid, so only the _attribute_ should be removed. If this ends up being a mandatory attribute, then the entire element can be removed as well, but it shouldn't be done here.

Reply:
Sorry my previous comment was a bit mangled. What I meant to say:

Given markup like this:

```html
<amp-img src="https://example.com/foo.jpg" width="" height="200" layout="responsive"></amp-img>
```

The layout is indeed invalid as there is no specified width, but as an attribute, the width itself is valid. Because there is no spec to validate the width, it will not be sanitized and the markup will be outputted.

Reply:
OK, well then the change to `sanitize_disallowed_attribute_values_in_node()` here should be reverted in favor of something like:

```diff
diff --git a/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php b/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
index 14e5cc00..a0c6b3bd 100644
--- a/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
+++ b/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
@@ -562,6 +562,11 @@ class AMP_Tag_And_Attribute_Sanitizer extends AMP_Base_Sanitizer {
 		// Fix any values for width & height that may be AMP incompatible.
 		$this->adapt_dimensions_to_layout( $node );
 
+		// If the layout is invalid, the node should be removed.
+		if ( ! $this->is_valid_layout( $tag_spec, $node ) && $this->remove_invalid_child( $node, [ 'code' => 'invalid_layout' ] ) ) {
+			return null;
+		}
+
 		// Identify any remaining disallowed attributes.
 		$disallowed_attributes = $this->get_disallowed_attributes_in_node( $node, $merged_attr_spec_list );
 
@@ -1044,12 +1049,6 @@ class AMP_Tag_And_Attribute_Sanitizer extends AMP_Base_Sanitizer {
 	private function sanitize_disallowed_attribute_values_in_node( DOMElement $node, $tag_spec, $attr_spec_list, $attributes_pending_removal ) {
 		$attrs_to_remove = [];
 
-		if ( ! $this->is_valid_layout( $tag_spec, $node ) ) {
-			// If the layout is invalid, the node should be removed.
-			$this->remove_node( $node );
-			return false;
-		}
-
 		foreach ( $attr_spec_list as $attr_name => $attr_val ) {
 			if ( isset( $attr_spec_list[ $attr_name ][ AMP_Rule_Spec::ALTERNATIVE_NAMES ] ) ) {
 				foreach ( $attr_spec_list[ $attr_name ][ AMP_Rule_Spec::ALTERNATIVE_NAMES ] as $attr_alt_name ) {
```